lib/Extension/LanguageServerCodeTransform/CodeAction/GenerateDecoratorProvider.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace Phpactor\Extension\LanguageServerCodeTransform\CodeAction;
+
+use Amp\CancellationToken;
+use Amp\Promise;
+use Phpactor\Extension\LanguageServerCodeTransform\LspCommand\GenerateDecoratorCommand;
+use Phpactor\LanguageServerProtocol\CodeAction;
+use Phpactor\LanguageServerProtocol\Command;
+use Phpactor\LanguageServerProtocol\Range;
+use Phpactor\LanguageServerProtocol\TextDocumentItem;
+use Phpactor\LanguageServer\Core\CodeAction\CodeActionProvider;
+use function Amp\call;
+
+class GenerateDecoratorProvider implements CodeActionProvider
+{
+    public const KIND = 'quickfix.generate_decorator';
+
+    public function kinds(): array
+    {
+        return [
+            self::KIND,
+        ];
+    }
+
+    public function provideActionsFor(TextDocumentItem $textDocument, Range $range, CancellationToken $cancel): Promise
+    {
+        return call(function () use ($textDocument) {
+            $this->generateDecoration->transform($textDocument);
```

Comment:
Don't mind this line, it's not working.

If we just take the first interface then this could be gone. Otherwise, here would be the place where we extract the interfaces and make them selectable via code actions.

Reply:
yeah, here we could have:

```php
$classes = $this->reflector->reflectClassesIn($textDocument->text);
if (count($classes) !== 1) {
    return [];
}
$class = $classes->first();
$interfaces = $class->interfaces();

if (count($interfaces) !== 1) {
    return [];
}

$methods = $class->methods();

if (count($methods) > 0) {
    return [];
}

if ($class->parent()) {
    return [];
}

// otherwise return code action
```

so in order to show code action:

- file must have one class
- class must have exactly one interface
- class must not extend anything
- class must not have any methods
- ...

can improve that as required, but I think it should be ok?

Reply:
Looks great, I have copied that over.